npm/src/clientfilemetadataupload/index.ts

```diff
@@ -50,51 +57,82 @@ export class ClientFileMetadataUpload {
     fileIds: string[],
     metadata: IFileMetadata[]
   ): Promise<ITdrFile[]> {
+    const metadataInputs: AddClientFileMetadataInput[] = []
     const files: ITdrFile[] = []
-    for (const element of metadata) {
-      let index = metadata.indexOf(element)
-      const fileId = fileIds[index]
-      const variables: AddClientFileMetadataMutationVariables = this.generateMutationVariables(
-        fileId,
-        element
-      )
+    this.generateClientFileData(fileIds, metadata).forEach(value => {
+      metadataInputs.push(value.metadataInput)
+      files.push(value.tdrFile)
+    })
 
+    const metadataBatches: AddClientFileMetadataInput[][] = this.createMetadataInputBatches(
+      metadataInputs
+    )
+
+    for (const metadataInputs of metadataBatches) {
+      const variables = { input: metadataInputs }
       const result = await this.client.mutation(
         AddClientFileMetadata,
         variables
       )
 
-      //Stop processing if an error is encountered
       if (result.errors) {
         throw Error(
-          "Add client file metadata failed for file " +
-            fileId +
-            ": " +
-            result.errors.toString()
+          "Add client file metadata failed: " + result.errors.toString()
         )
       }
-      const { file } = element
-      files.push({ fileId, file })
     }
+
     return files
   }
 
-  generateMutationVariables(
-    fileId: string,
-    metadata: IFileMetadata
-  ): AddClientFileMetadataMutationVariables {
-    return {
-      input: {
+  generateClientFileData(
+    fileIds: string[],
+    metadata: IFileMetadata[]
+  ): IClientFileData[] {
+    const clientFilesData: IClientFileData[] = []
+
+    metadata.forEach((value, index) => {
+      const fileId = fileIds[index]
```

Comment:
This is personal preference but you can do a scala style `map` here to convert from from one to the other. 
So instead of pushing to the clientFilesData variable, you can change the foreach to map and return the object. Something like:
```
return metadata.map((value, index) => {
      const fileId = fileIds[index]
     ................
     ................
      return {
        metadataInput,
        tdrFile
      }
    })

```

Reply:
Ah, that's pretty cool!